ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2835,232 @@ public sealed class Bot : IAsyncDisposable {
+
+				return;
+			}
+
+			HashSet<Steam.Asset> result = new HashSet<Steam.Asset>();
+
+			foreach (uint appID in appIDs) {
+				HashSet<Steam.Asset> itemsForAppBadges = await GetItemsForFullBadge(inventory, appID).ConfigureAwait(false);
+
+				result.UnionWith(itemsForAppBadges);
+			}
+
+			if (result.Count > 0) {
+				await Actions.SendInventory(inventory: result).ConfigureAwait(false);
+			}
+		}
+
+		private async Task<HashSet<Steam.Asset>> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, uint appID) {
+			byte requiredCards = await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false);
+
+			Dictionary<Steam.Asset.EType, HashSet<Steam.Asset>> inventoryForAppIDPerType = inventory.Where(item => item.RealAppID == appID).GroupBy(item => item.Type).ToDictionary(grouping => grouping.Key, grouping => grouping.ToHashSet());
+
+			HashSet<Steam.Asset> result = new HashSet<Steam.Asset>(0);
+
+			if (inventoryForAppIDPerType.TryGetValue(Steam.Asset.EType.TradingCard, out HashSet<Steam.Asset>? availableTradingCards)) {
+				result.UnionWith(GetItemsForFullBadge(availableTradingCards, requiredCards));
+			}
+
+			if (inventoryForAppIDPerType.TryGetValue(Steam.Asset.EType.FoilTradingCard, out HashSet<Steam.Asset>? availableFoilCards)) {
+				result.UnionWith(GetItemsForFullBadge(availableFoilCards, requiredCards));
+			}
+
+			return result;
+		}
+
+		private static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> availableItems, byte cardsPerBadge) {
+			Dictionary<ulong, List<Steam.Asset>> itemsPerClassId = availableItems.GroupBy(item => item.ClassID).ToDictionary(grouping => grouping.Key, grouping => grouping.OrderByDescending(item => item.Amount).ToList());
+
+			if (itemsPerClassId.Keys.Count != cardsPerBadge) {
+				// This can happen if cards are not tradable or we traded away a card between checking which badges are craftable and what our inventory state is
+				return new HashSet<Steam.Asset>(0);
+			}
+
+			// "maximum" as it could happen, that we need to send less sets than exist in our inventory. This only is possible if Valve ever decides to send us Assets with other amount than 1
+			// example: We have a set that is craftable 7 times. In our inventory we have item with class id 0 and amount 5 twice  and class id 1 4 times with amount 2. All other class ids are available seven times with an amount of 1 each.
+			// In this case we need to send 0 sets, but if we also have item with class id 0 and amount 1 once, we can send 6 sets
+			uint maximumSetCount = itemsPerClassId.Values.Select(items => items.Select(item => item.Amount).Aggregate((a, b) => a + b)).Min();
+
+			while (maximumSetCount > 0) {
+				HashSet<Steam.Asset>? itemsToGive = new HashSet<Steam.Asset>();
+
+				foreach (List<Steam.Asset> itemsOfClass in itemsPerClassId.Values) {
+					bool[] usedItems = itemsOfClass.Select(item => false).ToArray();
```

Comment:
Linq is fun, but could be simpler: `bool[] usedItems = new bool[itemsOfClass.Count];`